File: DSWaveformImage/DSWaveformImage/DSWaveformImageView.m
//
//  DSWaveformImageView.m
//  DSWaveformImage
//
//  Created by Dennis Schmidt on 04.11.13.
//  Copyright (c) 2013 dmrschmidt. All rights reserved.
//

#import "DSWaveformImageView.h"
#import "DSWaveformImage.h"

@implementation DSWaveformImageView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        _waveformColor = [UIColor darkGrayColor];
        _waveformStyle = DSWaveformStyleFull;
        _waveformPosition = DSWaveformPositionMiddle;
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)aDecoder {
    self = [super initWithCoder:aDecoder];
    if (self) {
        _waveformColor = [UIColor darkGrayColor];
        _waveformStyle = DSWaveformStyleFull;
        _waveformPosition = DSWaveformPositionMiddle;
    }
    return self;
}

- (void)setWaveformColor:(UIColor *)waveformColor {
    _waveformColor = waveformColor;
    [self updateWaveform];
}

- (void)setWaveformStyle:(DSWaveformStyle)waveformStyle {
    _waveformStyle = waveformStyle;
    [self updateWaveform];
}

- (void)setWaveformPosition:(DSWaveformPosition)waveformPosition {
    _waveformPosition = waveformPosition;
    [self updateWaveform];
}

- (void)setAudioURL:(NSURL *)audioURL {
    _audioURL = audioURL;
    [self updateWaveform];
}

#pragma mark - Private

- (void)updateWaveform {
    UIImage *image = nil;
    if (self.audioURL) {
        image = [DSWaveformImage waveformForAssetAtURL:self.audioURL
                                                 color:self.waveformColor
                                                  size:self.bounds.size
                                                 scale:[UIScreen mainScreen].scale
                                                 style:self.waveformStyle
                                              position:self.waveformPosition];
    }
    self.image = image;
}

@end

File: DSWaveformImage/DSWaveformImage/DSWaveformImageView.swift
import Foundation
import AVFoundation
import UIKit

public class DSWaveformImageView: UIImageView {
    fileprivate let waveformImageDrawer: DSWaveformImageDrawer

    public var waveformColor: UIColor {
        didSet { updateWaveform() }
    }

    public var waveformStyle: DSWaveformStyle {
        didSet { updateWaveform() }
    }

    public var waveformPosition: DSWaveformPosition {
        didSet { updateWaveform() }
    }

    public var waveformAudioURL: URL? {
        didSet { updateWaveform() }
    }

    override public init(frame: CGRect) {
        waveformColor = UIColor.darkGray
        waveformStyle = .gradient
        waveformPosition = .middle
        waveformImageDrawer = DSWaveformImageDrawer()
        super.init(frame: frame)
    }

    required public init?(coder aDecoder: NSCoder) {
        waveformColor = UIColor.darkGray
        waveformStyle = .gradient
        waveformPosition = .middle
        waveformImageDrawer = DSWaveformImageDrawer()
        super.init(coder: aDecoder)
    }

    override public func layoutSubviews() {
        super.layoutSubviews()
        updateWaveform()
    }
}

fileprivate extension DSWaveformImageView {
    func updateWaveform() {
        guard let audioURL = waveformAudioURL else { return }
        image = waveformImageDrawer.waveformImage(fromAudioAt: audioURL, color: waveformColor, style: waveformStyle,
                                                  position: waveformPosition, size: bounds.size,
                                                  scale: UIScreen.main.scale)
    }
}
